ml-agents/mlagents/trainers/buffer.py

```diff
@@ -12,321 +13,260 @@ class BufferException(UnityException):
+
+    def reset_agent(self) -> None:
+        """
+        Resets the AgentBuffer
+        """
+        for k in self.keys():
+            self[k].reset_field()
+        self.last_brain_info = None
+        self.last_take_action_outputs = None
 
     def __getitem__(self, key):
         if key not in self.keys():
-            self[key] = self.AgentBuffer()
-        return super(Buffer, self).__getitem__(key)
+            self[key] = self.AgentBufferField()
+        return super().__getitem__(key)
+
+    def check_length(self, key_list: List[str]) -> bool:
+        """
+        Some methods will require that some fields have the same length.
+        check_length will return true if the fields in key_list
+        have the same length.
+        :param key_list: The fields which length will be compared
+        """
+        if len(key_list) < 2:
+            return True
+        length = None
+        for key in key_list:
+            if key not in self.keys():
+                return False
+            if (length is not None) and (length != len(self[key])):
+                return False
+            length = len(self[key])
+        return True
+
+    def shuffle(self, sequence_length: int, key_list: List[str] = None) -> None:
+        """
+        Shuffles the fields in key_list in a consistent way: The reordering will
+        be the same across fields.
+        :param key_list: The fields that must be shuffled.
+        """
+        if key_list is None:
+            key_list = list(self.keys())
+        if not self.check_length(key_list):
+            raise BufferException(
+                "Unable to shuffle if the fields are not of same length"
+            )
+        s = np.arange(len(self[key_list[0]]) // sequence_length)
+        np.random.shuffle(s)
+        for key in key_list:
+            tmp: List[np.ndarray] = []
+            for i in s:
+                tmp += self[key][i * sequence_length : (i + 1) * sequence_length]
+            self[key][:] = tmp
+
+    def make_mini_batch(self, start: int, end: int) -> "AgentBuffer":
+        """
+        Creates a mini-batch from buffer.
+        :param start: Starting index of buffer.
+        :param end: Ending index of buffer.
+        :return: Dict of mini batch.
+        """
+        mini_batch = AgentBuffer()
+        for key in self:
+            mini_batch[key] = self[key][start:end]
+        return mini_batch
+
+    def sample_mini_batch(
+        self, batch_size: int, sequence_length: int = 1
+    ) -> "AgentBuffer":
+        """
+        Creates a mini-batch from a random start and end.
+        :param batch_size: number of elements to withdraw.
+        :param sequence_length: Length of sequences to sample.
+            Number of sequences to sample will be batch_size/sequence_length.
+        """
+        num_seq_to_sample = batch_size // sequence_length
+        mini_batch = AgentBuffer()
+        buff_len = len(next(iter(self.values())))
```

Comment:
nitL use self.num_experiences here?

Reply:
I think I finally got all of these calls out